test/cmdLineTests/jvmtitests/src/com/ibm/jvmti/tests/nestMatesRedefinition/nmr001.c

```
@@ -0,0 +1,81 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include <string.h>
+
+#include "jvmti_test.h"
+
+static agentEnv * env;
+
+
+jint JNICALL
+nmr001(agentEnv * agent_env, char * args)
+{
+	jvmtiError err;
+	jvmtiCapabilities capabilities;
+	JVMTI_ACCESS_FROM_AGENT(agent_env);
+
+	env = agent_env; 
+
+	memset(&capabilities, 0, sizeof(jvmtiCapabilities));
+	capabilities.can_redefine_classes = 1;
+	/* Following capability is required to enable FSD */
+	capabilities.can_generate_breakpoint_events = 1;
+	err = (*jvmti_env)->AddCapabilities(jvmti_env, &capabilities);
+	if (err != JVMTI_ERROR_NONE) {
+		error(env, err, "Failed to AddCapabilities");
+		return JNI_ERR;
+	}						
+
+	return JNI_OK;
+}
+
+jboolean JNICALL
+Java_com_ibm_jvmti_tests_nestMatesRedefinition_nmr001_redefineClass(JNIEnv * jni_env, jclass klass, jclass originalClass, jint classBytesSize, jbyteArray classBytes)
+{
+	JVMTI_ACCESS_FROM_AGENT(env);
+	jbyte * class_bytes;
+	char * classFileName = env->testArgs;
+	jvmtiClassDefinition classdef;
+	jvmtiError err; 
+	
+	err = (*jvmti_env)->Allocate(jvmti_env, classBytesSize, (unsigned char **) &class_bytes);
+	if (err != JVMTI_ERROR_NONE) {
+		error(env, err, "Unable to allocate temp buffer for the class file");
+		return JNI_FALSE;
+	}
+	
+	(*jni_env)->GetByteArrayRegion(jni_env, classBytes, 0, classBytesSize, class_bytes); 
+
+	classdef.class_bytes = (unsigned char *) class_bytes;
+	classdef.class_byte_count = classBytesSize;
+	classdef.klass = originalClass;
+	
+    err = (*jvmti_env)->RedefineClasses(jvmti_env, 1, &classdef);
+    (*jvmti_env)->Deallocate(jvmti_env, (unsigned char *) class_bytes);
+    if (err != JVMTI_ERROR_NONE) {
+    	error(env, err, "RedefineClasses failed");
+    	return JNI_FALSE;
+    }
+		
+	return JNI_TRUE;
```

can you return the error code instead, need to verify that it is failing the correct way

---

Will do.

It looks like the other jvmti test native functions only ever return `JNI_TRUE`/`JNI_FALSE` (jboolean) or `JNI_ERR`/`JNI_OK` (jint) - even in the other redefine classes tests, but print the JVMTI. Is there a likely reason that the other test cases don't usually bother checking the type of jvmti error?

---

Depends what the test is trying to do. In some cases all that matters is that the function succeeds or fails. In other cases the test is looking for a certain type of error. See mt001.java